### PR TITLE
Modify Timeout middleware to cooperate with RequeueExceptions middleware

### PR DESCRIPTION
**Pull request by vadim-moz**

### CODE REVIEW

File: lib/qless/middleware/timeout.rb [hunk 1]
```diff
@@ -15,29 +16,45 @@ module Middleware
...
+                error = JobTimedoutError.new(e.message)
+                error.set_backtrace(e.backtrace)
+                # The stalled connection to redis might be the cause of the timeout. We cannot rely
+                # on state of connection either (e.g., we might be in the middle of Redis call when
+                # timeout happend). To play it safe, we reconnect.
+                job.reconnect_to_redis
+                job.fail(*Qless.failure_formatter.format(job, error, []))
+                # Since we are leaving with bang (exit!), normal requeue logic does not work.
+                # Do it manually right here.
+                if self.is_a?(::Qless::Middleware::RequeueExceptions) && self.requeueable?(JobTimedoutError)
```
Comment (koos42) [inline lib/qless/middleware/timeout.rb]: It would seem cleaner if we could rely on the superposition of middlewares to handle retries here, instead of having the job check its own middlewares. Have you considered perhaps a middleware, like the RequeueMiddleware, that rescues exceptions coming through the middle ware stack and `kill!`s those that make it past the RequeueMiddleware?

File: lib/qless/middleware/requeue_exceptions.rb [hunk 2]
```diff
@@ -53,21 +53,29 @@ def on_requeue_callback
         @on_requeue_callback ||= DEFAULT_ON_REQUEUE_CALLBACK
       end
 
-      def around_perform(job)
-        super
-      rescue *requeueable_exceptions.keys => e
-        config = requeuable_exception_for(e)
+      def exception_handler(job, error)
```
Comment (myronmarston) [inline lib/qless/middleware/requeue_exceptions.rb]: WDYT about renaming this to `handle_exception`?  The method name `exception_handler` is a noun, which would suggest that it returns an object that is meant to be used to handle exceptions.  `handle_exception`, OTOH, as a verb is more clear that the method itself handles exceptions.

File: lib/qless/middleware/timeout.rb [hunk 3]
```diff
@@ -15,29 +16,45 @@ module Middleware
...
+        kernel_class = opts.fetch(:kernel_class, Kernel)
         module_eval do
           define_method :around_perform do |job|
-            timeout_value = yield job
-            if timeout_value <= 0
-              raise InvalidTimeoutError, "Timeout must be positive, but was #{timeout_value}"
-            end
+            timeout_seconds = yield job
+            if timeout_seconds.nil?
+              super(job)
```
Comment (myronmarston) [inline lib/qless/middleware/timeout.rb]: You can turn this `if/else` into an early return:

``` ruby
return super(job) if timeout_seconds.nil?

# the else logic goes here w/o a need to be nested under an `else`
```

I find that to be a bit cleaner.

File: lib/qless/middleware/timeout.rb (hunk 1 above)
Comment (vadim-moz) [inline lib/qless/middleware/timeout.rb]: Here is my design constrains:
* the both middlewares are in the same package (`Qless`), so they are delivered together.
* the timeout is very special case in the sense that it occurs outside of normal execution context. In particular it might have socket in weird state (including redis one). So we need to harden timeout processing logic for weird situations. In the same time, normal requeue does not need it.
* The (causing timeout) issue can be with Qless (redis) itself (and we saw it). So we need to protect against it. Part of it is `reconnect`. The Qless portion should be protected by secondary timeout, but I would rather implement is as a separate PR to avoid confusion.
* We want to avoid dependency of ordering requeue and timeout middleware.
* Thus some of the timeout logic must happen before, and some after requeue logic. 
* We want very loose coupling of two middlewares. So, while timeout middleware delegates to requeue one, the worker might be configured to be in any of four combinations.

Comment (myronmarston) [inline lib/qless/middleware/timeout.rb]: :+1: well said :).

File: lib/qless/middleware/timeout.rb (hunk 3 above)
Comment (vadim-moz) [inline lib/qless/middleware/timeout.rb]: I agree

File: lib/qless/middleware/requeue_exceptions.rb (hunk 2 above)
Comment (vadim-moz) [inline lib/qless/middleware/requeue_exceptions.rb]: it is rubyish tradition, which I slowly learn. Makes perfect sense.

**Comment by vadim-moz** [comment]
PTAL

File: spec/unit/middleware/requeue_exceptions_spec.rb [hunk 4]
```diff
@@ -16,8 +16,8 @@ def around_perform(job)
         end
       end
 
-      let(:container) do 
-        container = container_class.new 
+      let(:container) do
+        container = container_class.new
```
Comment (myronmarston) [inline spec/unit/middleware/requeue_exceptions_spec.rb]: I'm glad to see whitespace cleaned up, but in the future I would prefer to see whitespace cleanups in their own commit so they are not mixed with non-cosmetic code changes.

File: spec/unit/middleware/timeout_spec.rb [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        def around_perform(job)
+          @around_perform_call_counter = (@around_perform_call_counter || 0) + 1
+        end
+      end
+
+      let(:kernel_class) { class_double(Kernel) }
+
+      let(:job) {
+        instance_double(Qless::Job)
+      }
```
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: Why did you put this across 3 lines instead of one?

``` ruby
let(:job) { instance_double(Qless::Job) }
```

File: spec/unit/middleware/timeout_spec.rb [hunk 6]
```diff
@@ -0,0 +1,177 @@
+require 'spec_helper'
+require 'qless/job'
+require 'qless/middleware/timeout'
+
+module Qless
+  module Middleware
+    ::Rspec.describe Timeout do
```
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: The constant is `::RSpec`, not `::Rspec`.  I think RSpec 2 (which it looks like Qless is still using) provided `::Rspec` as a deprecated form of the constant for backwards compatibility.  It would be good to update to `::RSpec`.

Actually, it would be good to upgrade RSpec as well.  http://rspec.info/upgrading-from-rspec-2/ provides a step-by-step upgrade guide.  I'm happy to help (or even do it myself) if you like -- I've upgraded many other projects and can do so quickly and learning to upgrade RSpec isn't a particularly useful skill to gain for others.

File: spec/unit/middleware/timeout_spec.rb [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+      class JobClass
+        attr_accessor :perform
```
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: Is this `perform` used anywhere?  As far as I can see, it's not.  Anyhow, Qless expects the `perform` method to be a class method on the job class, but here it is an instance method.  I guess it does not matter since this is in isolation as a unit test so it's not qless calling it but it is a bit confusing that it does not conform to the Qless job protocol but appears to be trying to.

File: spec/unit/middleware/timeout_spec.rb [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+          include Qless::Middleware::Timeout.new(timeout_class: timeout_class,
+              kernel_class: kernel_class) { timeout_seconds }
+        end.new
+      end
+
+      before {
+        allow(job).to receive(:reconnect_to_redis)
+        allow(job).to receive(:klass_name)
+        allow(job).to receive(:fail)
+        allow(kernel_class).to receive(:exit!)
```
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: I think the specs here could benefit from using `instance_double(...).as_null_object` or `instance_spy(...)` (the latter is only available on recent versions of RSpec 3.  Then you wouldn't need to explicitly stub these methods, and in the specs below you could stick with the familiar arrange-act-assert pattern by using `expect(...).to have_received(...)` rather than ahead-of-time `expect(...).to receive(...)`.

If you want to make that change it would almost certainly help to upgrade to RSpec 3 as it's support for using doubles as spies is much better.

File: spec/unit/middleware/timeout_spec.rb [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      context 'when timeout is not nil' do
+        it 'aborts with a clear error when given a non-numeric timeout' do
+          worker = make_worker(TriggeredTimeout, "123", kernel_class)
+
+          expect {
+            worker.around_perform job
+          }.to raise_error(Qless::InvalidTimeoutError)
+        end
+
+        it 'aborts with a clear error when given a non-positive timeout' do
```
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: Is it worth making another spec for the `0` case?  This doc string states that `0` should have the same behavior as the `-1` below but there is no example demonstrating that.  (Not a big deal to me either way but perhaps worth doing).

File: spec/unit/middleware/timeout_spec.rb [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+              worker.extend(RequeueExceptions)
+              expect(worker).not_to receive(:handle_exception).with(job, anything)
+
+              expect {
+                worker.around_perform job
+              }.not_to raise_error
+            end
+          end
+
+          it 'terminates process' do
```
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: I think this doc string could benefit from some additional explanation:

``` ruby
it 'terminates the process so it is not left in an inconsistent state (since `Timeout` can do that)'
```

**Comment by myronmarston** [comment]
LGTM. I don't consider any of my feedback worthy of blocking a merge, so address (or not) whichever comments you like.

File: spec/unit/middleware/requeue_exceptions_spec.rb (hunk 4 above)
Comment (vadim-moz) [inline spec/unit/middleware/requeue_exceptions_spec.rb]: It is a side effect. I updated file, and the editor did cleanup.

There is three setting to editor (not all editors support all of them): (a) ignore tailing whitespaces, (b) kill them, and (c) kill only in modified lines. There are also different team policies. I understand your frustration, and would propose to a global whitespace cleanup with lint check afterwards. Should we take it to retrospective?

File: spec/unit/middleware/timeout_spec.rb (hunk 6 above)
Comment (vadim-moz) [inline spec/unit/middleware/timeout_spec.rb]: I will try to do upgrade as a separate PR.

File: spec/unit/middleware/timeout_spec.rb (hunk 7 above)
Comment (vadim-moz) [inline spec/unit/middleware/timeout_spec.rb]: leftover of copy-paste

File: spec/unit/middleware/timeout_spec.rb (hunk 8 above)
Comment (vadim-moz) [inline spec/unit/middleware/timeout_spec.rb]: Apparently, in current version of RSpec the order cannot be changed. It annoys me :-( but stimulates RSpec migration

File: spec/unit/middleware/timeout_spec.rb (hunk 9 above)
Comment (vadim-moz) [inline spec/unit/middleware/timeout_spec.rb]: I think it is of low value. 0 is one of the non-positive values, and we cannot write test for each one. Grunted, it is a boundary case.

This clearly rings similar to our discussion in your PR a few days ago. The difference for me is that their comparison was calculated (4 days from now -- does it mean that 4th day is included or excluded), while here it is very explicit. 

That said, I am ready to add an example upon confirmation.

File: spec/unit/middleware/timeout_spec.rb (hunk 5 above)
Comment (vadim-moz) [inline spec/unit/middleware/timeout_spec.rb]: a vestige of debug. Fixed.

File: spec/unit/middleware/timeout_spec.rb (hunk 9 above)
Comment (myronmarston) [inline spec/unit/middleware/timeout_spec.rb]: No need.

File: spec/unit/middleware/requeue_exceptions_spec.rb (hunk 4 above)
Comment (myronmarston) [inline spec/unit/middleware/requeue_exceptions_spec.rb]: I'm actually not frustrated :).

As far as I know, everyone the team is on board with not committing trailing whitespace (particularly since we enforce it in some of the other repos).  Given that, I'm not convinced it would be worthwhile to spend retrospective time on it.  A global whitespace cleanup with lint check afterwards is fine, IMO.

> There is three setting to editor (not all editors support all of them): (a) ignore tailing whitespaces, (b) kill them, and (c) kill only in modified lines.

FWIW, There's a 4th that I use: a keybinding that allows me to strip whitespace when I want to.  I use it whenever I noticed trailing whitespace (which my editor highlights) but make sure to put it in its own commit.

File: spec/unit/middleware/timeout_spec.rb (hunk 10 above)
Comment (vadim-moz) [inline spec/unit/middleware/timeout_spec.rb]: Thank you. Done

**Comment by vadim-moz** [comment]
Thank you!
